### PR TITLE
Validate a TxDefaultProfile.

### PR DESCRIPTION
**Pull request by gberardi-pillar**
These changes apply to K01.FR.52 and K01.FR.53.

Given a TxDefaultProfile on an EVSE or on the charging station as a whole, we can validate whether a new TxDefaultProfile is allowed.

### CODE REVIEW

**Comment by gberardi-pillar** [comment]
@shankari Please review these changes.

**Comment by gberardi-pillar** [comment]
@shankari In case you missed my previous comment, can you please review this PR?

**Review by shankari — CHANGES_REQUESTED** [review]
- I think that some additional documentation would be helpful
- there is a static code analysis error in here

File: include/ocpp/v201/evse.hpp
Comment (gberardi-pillar) [reply include/ocpp/v201/evse.hpp:76]: An EVSE ID > 0 is a specific EVSE, but 0 is meant to be a station-wide value to apply to all EVSEs for that station.

The spec does not explicitly say so in so many words, but there are a few places in which an evseid = 0 means it applies to all EVSEs: "For TxDefaultProfile an evseId=0 applies the profile to each individual evse. For ChargingStationMaxProfile and ChargingStationExternalConstraints an evseId=0 contains an overal limit for the whole Charging Station."

Comment (gberardi-pillar) [reply include/ocpp/v201/evse.hpp]: I have updated the comment with the above quote from the spec.

**Review by Pietfried — CHANGES_REQUESTED** [review]
It is not intended that an Evse instance is initialized with ID 0. Ids of Evses must start from 1 upwards as required by 7.1 EVSE Numbering in OCPP2.0.1 Part 1 - Architecture & Topology. This requires a redesign of the `validate_tx_default_profile` function signature, where I think the `Evse` argument could simply be replaced with int32_t `evse_id`.

The Evse class does there not require a member function `is_station_wide`. 

Additionally, a more general note: Im not sure if the member `std::vector<EvseProfile> charging_profiles` of the `std::vector<EvseProfile> charging_profiles;` is complex enough to represent more complex combinations of profiles. The OCPP1.6 implementation makes use of two different std::map per EVSE (in 1.6 Evse and Connector is equivilant):
https://github.com/EVerest/libocpp/blob/main/include/ocpp/v16/connector.hpp#L20

**Review by Pietfried — CHANGES_REQUESTED** [review]
lgtm, except for some minor function doc updates and the argument for `validate_tx_default_profile`

File: include/ocpp/v201/smart_charging.hpp
Comment (gberardi-pillar) [reply include/ocpp/v201/smart_charging.hpp]: We've been treating the existence of an Evse as a specific instance and the lack of one as indicating ID = 0. I did this work to follow the pattern we have established elsewhere. 

But I can use an ID and check if it is 0 instead.

File: include/ocpp/v201/smart_charging.hpp
Comment (gberardi-pillar) [reply include/ocpp/v201/smart_charging.hpp:74]: Done.

**Review by Pietfried — APPROVED** [review]
:+1: 
This needs a rebase since 
https://github.com/EVerest/libocpp/pull/534 has been merged

File: include/ocpp/v201/smart_charging.hpp
Comment (marcemmers) [inline include/ocpp/v201/smart_charging.hpp:47]: Would an `std::map` with the evse_id and a vector of charging profiles make sense here? That way we don't need the specific type for `EvseProfile` and we can search a bit quicker too.

File: lib/ocpp/v201/smart_charging.cpp
Comment (marcemmers) [inline lib/ocpp/v201/smart_charging.cpp:132]: If there are evse specific default profiles, wouldn't it make sense to filter on them as well? Now you return all profiles for all evse's. I would expect a test to expose this issue as well.

File: lib/ocpp/v201/smart_charging.cpp
Comment (marcemmers) [inline lib/ocpp/v201/smart_charging.cpp:134]: If it is possible I would suggest to use the range based for loop here and in the other for loops. More readable and less error prone.

**Review by marcemmers — CHANGES_REQUESTED** [review]
Sorry I am a bit late to the party here but I do have some suggestions/questions.

File: lib/ocpp/v201/smart_charging.cpp
Comment (gberardi-pillar) [reply lib/ocpp/v201/smart_charging.cpp:132]: I'm unclear on what issue you are referring to. Are you saying that it shouldn't return all profiles for all EVSEs?

Otherwise, while you might be right that it might make sense to eventually filter based on EVSE, it wasn't necessary for this code, and I did not want to write code for something I might anticipate a need for, increasing the size of the PR and distracting from the focus of it.

Unless I misunderstand what you mean?

Comment (marcemmers) [reply lib/ocpp/v201/smart_charging.cpp:132]: It could very well be that I misunderstood the intended functionality here.

What I meant is that if you add a TxDefaultProfile for EVSE 1 and then try to add a TxDefaultProfile for EVSE 2 I would expect that to be possible. Interpreting the code here I expect you to get a `ProfileValidationResultEnum::DuplicateTxDefaultProfileFound` error instead.

Comment (gberardi-pillar) [reply lib/ocpp/v201/smart_charging.cpp:132]: Ah, I see what you are saying now. You're right. I can add a test case for that case.

**Comment by gberardi-pillar** [comment]
Put PR back into draft due to still waiting for @shankari to re-review.

File: lib/ocpp/v201/smart_charging.cpp
Comment (gberardi-pillar) [reply lib/ocpp/v201/smart_charging.cpp:132]: Turned out that it was already going to pass.

Comment (marcemmers) [reply lib/ocpp/v201/smart_charging.cpp:132]: I guess it now passes since you are adding another profile with a different `ChargingProfile.id`. This makes sense now I've read this part of the spec because we can never have multiple profiles with the same ID according to usecase K01.FR.05. I did not know that before.

File: lib/ocpp/v201/smart_charging.cpp
Comment (gberardi-pillar) [reply lib/ocpp/v201/smart_charging.cpp]: Done.

Comment (marcemmers) [reply lib/ocpp/v201/smart_charging.cpp]: Looks quite a lot cleaner :+1:

File: include/ocpp/v201/smart_charging.hpp
Comment (gberardi-pillar) [reply include/ocpp/v201/smart_charging.hpp]: Ah, yes, you're correct.

File: include/ocpp/v201/smart_charging.hpp
Comment (gberardi-pillar) [reply include/ocpp/v201/smart_charging.hpp]: I am assuming you saw the response to this same question on the EVSE validation PR.

File: include/ocpp/v201/evse.hpp
Comment (shankari) [reply include/ocpp/v201/evse.hpp]: Please also add this assumption to the OCPP 201 document in https://github.com/EVerest/libocpp/blob/main/doc/ocpp_201_status.md

File: include/ocpp/v201/smart_charging.hpp
Comment (shankari) [reply include/ocpp/v201/smart_charging.hpp]: Yes. I am unable to resolve my comments, though. Feel free to resolve this if you have write acces.

**Review by shankari — APPROVED** [review]
@gberardi-pillar the changes look fine to me. However, I would suggest that, in the future, you have more detailed commit messages than "Review fix". A short description of what part of the review is being addressed by the commit would be helpful, along with a summary of the changes. You could even add a link to the commit to make it easier for future maintainers to easily find the review without having to search through all the comments in the PR

**Review by marcemmers — APPROVED** [review]

**Comment by gberardi-pillar** [comment]
@Pietfried @marcemmers Is this PR ready to be merged?

**Comment by marcemmers** [comment]
> @Pietfried @marcemmers Is this PR ready to be merged?

Yes, go for it :+1: 
Might be good to do a rebase before merging. Don't expect any issues there.

File: include/ocpp/v201/evse.hpp
Comment (gberardi-pillar) [reply include/ocpp/v201/evse.hpp]: This code doesn't exist in the PR anymore.

**Comment by drmrd** [comment]
Contributes to #361

**Pull request merged**
